sklearn/metrics/classification.py

```diff
@@ -391,72 +392,273 @@ def jaccard_similarity_score(y_true, y_pred, normalize=True,
+                                          sample_weight=weights)
+            if average == 'micro':
+                pred_or_true = np.array([pred_or_true.sum()])
+                pred_and_true = np.array([pred_and_true.sum()])
+
             score = pred_and_true / pred_or_true
             score[pred_or_true == 0.0] = 1.0
+
+            if average is not None:
+                if normalize is False:
+                    if class_weight is not None:
+                        score = np.dot(score, class_weight)
+                    else:
+                        score = score.sum()
+                else:
+                    score = np.average(score, weights=class_weight)
+            return score
+    elif average == 'samples':
+        raise ValueError("Sample-based jaccard similarity score is "
+                         "not meaningful outside multilabel "
+                         "classification. See the accuracy_score instead.")
     else:
-        score = y_true == y_pred
+        le = LabelEncoder()
+        le.fit(labels)
+        y_true = le.transform(y_true)
+        y_pred = le.transform(y_pred)
 
-    return _weighted_sum(score, sample_weight, normalize)
+        labels = le.transform(labels)[:n_labels]
+        # use 'np.in1d' instead of 'np.isin' (unavailable in version < 1.13.0)
+        indices = np.where(np.in1d(y_true, labels, assume_unique=False,
+                                   invert=False)
+                           + np.in1d(y_pred, labels, assume_unique=False,
+                                     invert=False))[0]
+
+        y_true = y_true[indices]
+        y_pred = y_pred[indices]
+        tp = y_true == y_pred
+        tp_bins = y_true[tp]
+        if sample_weight is not None:
+            sample_weight = np.array(sample_weight)[indices]
+            tp_bins_weights = np.asarray(sample_weight)[tp]
+        else:
+            tp_bins_weights = None
+
+        if len(tp_bins):
+            tp_sum = np.bincount(tp_bins, weights=tp_bins_weights,
+                                 minlength=len(labels))[labels]
+            true_sum = np.bincount(y_true, weights=sample_weight,
+                                   minlength=len(labels))[labels]
+            pred_sum = np.bincount(y_pred, weights=sample_weight,
+                                   minlength=len(labels))[labels]
+        else:
+            tp_sum = np.zeros(len(labels))
+            if len(y_true):
+                true_sum = np.bincount(y_true, weights=sample_weight,
+                                       minlength=len(labels))[labels]
+            else:
+                true_sum = np.zeros(len(labels))
+            if len(y_pred):
+                pred_sum = np.bincount(y_pred, weights=sample_weight,
+                                       minlength=len(labels))[labels]
+            else:
+                pred_sum = np.zeros(len(labels))
+
+        with np.errstate(divide='ignore', invalid='ignore'):
```

Comment:
Use as small a context as possible for this: only put errstate around the division.